### ISSUE TITLE
Is there no backwards compatablility for yolov5?

### ISSUE THREAD

**Issue by guzman109** [ultralytics/yolov5#13526]
### Search before asking

- [x] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I'm using the ultralytics library and loading the model via torch hub. Looks like NMS from ultralytics is different than the NMS from the yolov5 repo and I am getting completely different results between using both models as well.

Is there no backwards compatibility?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @guzman109, thank you for your interest in YOLOv5 🚀! Please visit our ⭐️ [Tutorials](https://docs.ultralytics.com/yolov5/) to get started, where you can find quickstart guides for simple tasks like [Custom Data Training](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data/) all the way to advanced concepts like [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution/).

If this is a 🐛 Bug Report, please provide a **minimum reproducible example** to help us debug it. Specifically, for your scenario, please include:
1. The exact code snippet you're using to load the model via `torch.hub` and perform inference.
2. Details of the YOLOv5 version(s) you're comparing, and how you're loading the models in both cases.
3. Examples of the input image(s) and the resulting outputs you're observing for comparison.

If this is a custom model or compatibility ❓ Question, please provide as much information as possible, including the details of the models or libraries you're using. Additionally, consider ensuring that all setups are aligned with our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

## Requirements

[**Python>=3.8.0**](https://www.python.org/) with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) installed including [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). To get started:
```bash
git clone https://github.com/ultralytics/yolov5  # clone
cd yolov5
pip install -r requirements.txt  # install
```

## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml"><img src="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml/badge.svg" alt="YOLOv5 CI"></a>

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are currently passing. CI tests verify correct operation of YOLOv5 [training](https://github.com/ultralytics/yolov5/blob/master/train.py), [validation](https://github.com/ultralytics/yolov5/blob/master/val.py), [inference](https://github.com/ultralytics/yolov5/blob/master/detect.py), [export](https://github.com/ultralytics/yolov5/blob/master/export.py) and [benchmarks](https://github.com/ultralytics/yolov5/blob/master/benchmarks.py) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response to help guide you with helpful resources 💡. An Ultralytics engineer will review your issue and assist you further shortly.

**Comment by pderrenger** [comment]
@guzman109 yOLOv5 maintains backward compatibility for inference when using the **exact same model version** across both torch hub and local code. For optimal results:

1. **Specify exact model version** in torch.hub.load():
```python
model = torch.hub.load('ultralytics/yolov5', 'yolov5s', version='v6.0')  # specify version tag
```
2. Verify you're using matching:
   - Model architecture (v6.0+ has updated layers)
   - Postprocessing (NMS implementation in `utils.general.non_max_suppression`)
   - Package versions (`pip install -U ultralytics`)

The [YOLOv5 v6.0 release notes](https://ultralytics.com/blog/yolov5-v6-0-is-here) detail architectural changes that might affect compatibility. If issues persist, please share a minimum reproducible example per our [bug report guidelines](https://github.com/ultralytics/yolov5/issues/4032).

**Comment by guzman109** [comment]
So the output changed at some point to match the other yolo models? The non_max_suppression input has a different shape than older version of the model.

Also, is scale_boxes from utils/general.py the same as scale_boxes or scale_coords in ultralytics module?

**Comment by pderrenger** [comment]
@guzman109 yOLOv5 v6.0+ introduced architectural updates that changed output dimensions and postprocessing steps to align with newer YOLO versions. For compatibility:

1. **NMS Input Shape**: Use matching model and code versions (v6.0+ models require v6.0+ code). Older versions have different output shapes as detailed in the [YOLOv5 v6.0 release notes](https://www.ultralytics.com/blog/yolov5-v6-0-is-here).

2. **Scaling Functions**: `scale_coords` was renamed to `scale_boxes` in newer versions but serves the same purpose. Use the function matching your code version from the current [YOLOv5 utils](https://github.com/ultralytics/yolov5/blob/master/utils/general.py).

For specific version behaviors, see the [Non-Maximum Suppression glossary](https://www.ultralytics.com/glossary/non-maximum-suppression-nms) and ensure your `ultralytics` package is updated (`pip install -U ultralytics`). Let us know if you need version-specific details.
